File: 1-general-examples/code-snippets/python/code_blocks.py
import requests
import json
import urllib3
import argparse
import os
import getpass

# Parse command line arguments
parser = argparse.ArgumentParser(description='Some explanation of what the python script does')
parser.add_argument('-u', '--username', dest='username', type=str, help='Username to login to Control-M/Enterprise Manager')
parser.add_argument('-pf', '--pwfile', dest='pwfile', type=str, help='The file that contains the password to login to Control-M/Enterprise Manager')
parser.add_argument('-h', '--host', dest='host', type=str, help='Control-M/Enterprise Manager hostname')
parser.add_argument('-i', '--insecure', dest='insecure', action='store_const', const=True, help='Disable SSL Certification Verification')
args = parser.parse_args()

# Error handling
# 1. Supress insecure request warning
if args.insecure:
    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
verifycert = not args.insecure

if args.pwfile is not None:
    if not os.path.exists(args.pwfile):
        print("Unable to read the specified file, %s, or file does not exist." % args.pwfile)
    else:
        with open(args.pwfile, "r") as f:
            passwd = f.read()
            passwd = passwd.rstrip(' \t\r\n\0')
else:
    passwd = getpass.getpass("Password: ")

# Login to the API
try: # start a try/except to catch specific expections related to the http connection
    r = requests.post('https://'+host+':8443/automation-api/session/login', 
            json={"password": passwd, "username": args.username},
            verify=verifycert)
except requests.exceptions.ConnectTimeout as err: # If the connection times out waiting for the response from the server
    print("Connecting to Automation API REST Server failed with error: " + str(err))
    exit(1)
except requests.exceptions.SSLError as err: # If there is an SSL error.
    print("Connecting to Automation API REST Server failed with error: " + str(err))
    print('INFO: If using a Self Signed Certificate use the -i flag to disable cert verification or \ # If you have created a flag or config file option to disable tls/ssl validation,
            add the certificate to this systems trusted CA store') #  this is a good place to hint to the user that they may want to use it
    exit(1)
except requests.exceptions.HTTPError as err: # Catch any general HTTP errors like 500/400 responces
    print("Connecting to Automation API REST Server failed with error: " + str(err))
    exit(1)

if 'errors' in json.loads(r.text): # Check if the response contains an object named "error"
    print("Unable to login!") # if so, print a reason able message based on the action that was just attempted
    print(json.dumps(json.loads(r.text)['errors'][0]['message'])) # then print the value of the message key for the json object(s) inside of the errors array
    quit(1) # quit/exit the program if desired. (Retry logic could be implemented here instead of a quit())
else:
    # Logic/Actions on the result, json.loads(r.text), go here
    if 'token' in json.loads(r.text):
        token = json.loads(r.text)['token']
        passwd = None
        del passwd
    else:
        print("Could not retreive token")
        quit(1)

File: 1-general-examples/code-snippets/java/code_blocks.java

import java.io.IOException;
import java.nio.file.*;
import java.util.concurrent.Callable;

import org.json.JSONObject;

import kong.unirest.*;
import picocli.CommandLine;
import picocli.CommandLine.*;

@Command(name="code_blocks")
class code_blocks implements Callable<Integer> {

	@Option(names = {"-u"}, required=true, description="Username to login to Control-M/Enterprise Manager")
	private String user;

	private String pass;

	@Option(names = {"-pf"}, description="The file that contains the password to login to Control-M/Enterprise Manager")
	private String pass_file = null;

	@Option(names = {"-h"}, required=true, description="Control-M/Enterprise Manager Hostname")
	private String host;

	@Option(names = {"-i"}, description="Disable SSL Certificate Validation")
	private boolean insecure = false;

	@Option(names = {"--help"}, usageHelp=true, description="show this help message and exit")
	private boolean help = false;

	public static void main(String[] args) {
		int exitCode = new CommandLine(new code_blocks()).execute(args);
		System.exit(exitCode);
	}

	@Override
	public Integer call() {
		// Ignore SSL verification if we're told to do so
		Unirest.config().verifySsl(!insecure);

		// Retrieve the password
		if (pass_file != null) {
			try {
				pass = new String(Files.readAllBytes(Paths.get(pass_file)), "UTF-8").trim();
			} catch (IOException e) {
				System.err.printf("Unable to read password file: %s\n", pass_file);
				System.err.println(e.getMessage());
				Unirest.shutDown();
				return 1;
			}
		} else {
			System.out.print("Password: ");
			pass = new String(System.console().readPassword());
		}

		HttpResponse<JsonNode> response;

		// Login
		try {
			response = Unirest.post(String.format("https://%s/automation-api/session/login", host))
				.header("Content-Type", "application/json")
				.body(String.format("{\"username\":\"%s\", \"password\":\"%s\"}", user, pass))
				.asJson();
		} catch (UnirestException e) {
			System.err.println(e.getMessage());
			Unirest.shutDown();
			return 1;
		}

		// Handle error
		if (!response.isSuccess()) {
			for (Object m: response.getBody().getObject().getJSONArray("errors"))
				System.err.println(((JSONObject)m).getString("message"));
			Unirest.shutDown();
			return 1;
		}

		String token = response.getBody().getObject().getString("token");

		System.out.println(token);

		// Logout
		try {
			response = Unirest.post(String.format("https://%s/automation-api/session/logout", host))
					.header("Content-Type", "application/json")
					.header("Authorization", String.format("Bearer %s", token))
					.asJson();
		} catch (UnirestException e) {
			System.err.println(e.getMessage());
			Unirest.shutDown();
			return 1;
		}

		// Handle error
		if (!response.isSuccess()) {
			for (Object m: response.getBody().getObject().getJSONArray("errors"))
				System.err.println(((JSONObject)m).getString("message"));
			Unirest.shutDown();
			return 1;
		}

		Unirest.shutDown();
		return 0;
	}
}
